### PR TITLE
Timezone feature

### PR DESCRIPTION
**Pull request by csduarte**
#### Summary
1. Supply `SupportedTimezones` via config to clients
2. Update PatchUser logic to update Timezones

https://github.com/mattermost/mattermost-webapp/pull/696
https://github.com/mattermost/mattermost-mobile/pull/1456

#### Ticket Link
[PLT-541 Add a "Time zone" field to account settings - Mattermost](https://mattermost.atlassian.net/browse/PLT-541)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)
- [x] Has UI changes

### CODE REVIEW

**Comment by miguelespinoza** [comment]
webapp PR for reference: https://github.com/mattermost/mattermost-webapp/pull/696

**Comment by miguelespinoza** [comment]
Could I get details on what's failing? I don't have access to the build server

**Comment by esethna** [comment]
@hmhealey could you help provide some details on the build failure when you have a chance?

**Comment by hmhealey** [comment]
You've got some unit tests failing
```
--- FAIL: TestGetClientProperties (0.37s)
	general_test.go:20: 
```

```
--- FAIL: TestGetOldClientConfig (0.80s)
	system_test.go:171: config not returned correctly
```

```
--- FAIL: TestPatchUser (0.95s)
	user_test.go:1027: Timezone did not update properly
```

**Comment by miguelespinoza** [comment]
Thanks @hmhealey. I'm working on fixing those test along with making a refactor to not have to worry about changing map[string]string to map[string]interface

**Comment by jwilander** [comment]
Skipping to dev review as the server PR isn't PM-testable

File: store/sqlstore/upgrade.go [hunk 1]
```diff
@@ -350,6 +351,17 @@ func UpgradeDatabaseToVersion47(sqlStore SqlStore) {
...
+		defaultTimezone := make(map[string]string)
+		defaultTimezone["useAutomaticTimezone"] = "true"
+		defaultTimezone["automaticTimezone"] = ""
+		defaultTimezone["manualTimezone"] = ""
+
+		defaultTimezoneValue, err := json.Marshal(defaultTimezone)
+		if err != nil {
+			l4g.Critical(err)
+		}
+		sqlStore.CreateColumnIfNotExists("Users", "Timezone", "varchar(150)", "varchar(150)", string(defaultTimezoneValue))
```
Comment (jwilander) [inline store/sqlstore/upgrade.go]: I think you'll also need to set the column size limit here https://github.com/mattermost/mattermost-server/blob/master/store/sqlstore/user_store.go#L81

Is 150 arbitrary or is that close to the maximum length of the field using the longest timezone, etc.?

File: utils/config.go [hunk 2]
```diff
@@ -448,6 +452,8 @@ func GenerateClientConfig(c *model.Config, diagnosticId string, license *model.L
 	props["DiagnosticsEnabled"] = strconv.FormatBool(*c.LogSettings.EnableDiagnostics)
 
 	props["PluginsEnabled"] = strconv.FormatBool(*c.PluginSettings.Enable)
+	props["EnableTimezoneSelection"] = strconv.FormatBool(*c.DisplaySettings.EnableTimezoneSelection)
+	props["SupportedTimezones"] = strings.Join(c.SupportedTimezones, ",")
```
Comment (jwilander) [inline utils/config.go]: This is going to make the client config a lot larger and could potentially slow down first load as a result. The client only needs to know this when the user is selecting a timezone, right? If so I wonder if it's better to isolate this into it's own REST API endpoint that gets called when the user opens their account settings to the timezone page

File: utils/config.go [hunk 3]
```diff
@@ -284,7 +284,7 @@ func EnsureConfigFile(fileName string) (string, error) {
 // LoadConfig will try to search around for the corresponding config file.  It will search
 // /tmp/fileName then attempt ./config/fileName, then ../config/fileName and last it will look at
 // fileName.
-func LoadConfig(fileName string) (config *model.Config, configPath string, appErr *model.AppError) {
+func LoadConfig(fileName string, isTimezoneConfig bool) (config *model.Config, configPath string, appErr *model.AppError) {
```
Comment (lieut-data) [inline utils/config.go]: Instead of requiring `LoadConfig(fileName, false)` everywhere, would it make sense to instead expose a `LoadTimezoneConfig`, and instead factor out a `loadConfigFile` for the common code?

File: config/default.json [hunk 4]
```diff
@@ -211,6 +211,9 @@
         "AllowCustomThemes": true,
         "AllowedThemes": []
     },
+    "DisplaySettings": {
+        "EnableTimezoneSelection": false
```
Comment (hmhealey) [inline config/default.json]: 👍 We've needed a section like this for such a long time. We keep jamming this type of setting into ServiceSettings even though it there's so much random stuff in there now that it makes sense to separate this out

File: store/sqlstore/upgrade.go [hunk 5]
```diff
@@ -350,6 +351,17 @@ func UpgradeDatabaseToVersion47(sqlStore SqlStore) {
 		sqlStore.AlterColumnTypeIfExists("OAuthAuthData", "State", "varchar(1024)", "varchar(1024)")
 		sqlStore.RemoveColumnIfExists("ChannelMemberHistory", "Email")
 		sqlStore.RemoveColumnIfExists("ChannelMemberHistory", "Username")
+
+		defaultTimezone := make(map[string]string)
```
Comment (hmhealey) [inline store/sqlstore/upgrade.go]: I think it might make sense to move the code that makes this default timezone map into `model/user.go` in case we need to construct an object like that elsewhere.

File: model/config.go [hunk 6]
```diff
@@ -1708,6 +1721,7 @@ type Config struct {
 	MessageExportSettings MessageExportSettings
 	JobSettings           JobSettings
 	PluginSettings        PluginSettings
+	SupportedTimezones    SupportedTimezones
```
Comment (hmhealey) [inline model/config.go]: This feels like something that should either be hardcoded or grabbed from the OS, not something that anyone would actually configure

File: store/sqlstore/upgrade.go (hunk 1 above)
Comment (miguelespinoza) [reply store/sqlstore/upgrade.go]: @jwilander, 150 refers to the closest maximum length with some padding (maybe like 6 char additional padding)

Could you tell me what's the difference between `varchar(150)` and `table.ColMap("Position").SetMaxSize(128)`

File: utils/config.go (hunk 2 above)
Comment (miguelespinoza) [reply utils/config.go]: It could potentially cause a slow down on first load. 

let me get some metrics to see what the difference makes with this change

File: utils/config.go (hunk 3 above)
Comment (miguelespinoza) [reply utils/config.go]: Sure, it wasn't that drastic of a refactor so I decided to add a bool instead.

not sure what your thoughts are on that. Only common code would be:

```
if fileName != filepath.Base(fileName) {
 		configPath = fileName
 	} else {
 		if path, err := EnsureConfigFile(fileName); err != nil {
 			appErr = model.NewAppError("LoadConfig", "utils.config.load_config.opening.panic", map[string]interface{}{"Filename": fileName, "Error": err.Error()}, "", 0)
 			return
 		} else {
 			configPath = path
 		}
 	}
 
 	config, err := ReadConfigFile(configPath, true)
 	if err != nil {
 		appErr = model.NewAppError("LoadConfig", "utils.config.load_config.decoding.panic", map[string]interface{}{"Filename": fileName, "Error": err.Error()}, "", 0)
 		return
 	}
```

Yeh, I can do that refactor

File: model/config.go (hunk 6 above)
Comment (miguelespinoza) [reply model/config.go]: It's pulling from a seperate file timezone.json so in a sense it is hardcoded.

pulling from OS would be problematic when you have the server on clusters.  Therefore, we took this apporach, of having a json for all timezones.

Something setting up the server does have the liberty to add only certain timezone.  Not sure why, but possible. Maybe all employees are in one location/region 🤷‍♂️

File: store/sqlstore/upgrade.go (hunk 5 above)
Comment (miguelespinoza) [reply store/sqlstore/upgrade.go]: Yeh, this would clean it up to be used elsewhere if needed.

File: utils/config.go (hunk 2 above)
Comment (miguelespinoza) [reply utils/config.go]: @jwilander I'll work on getting this refactored into an endpoint.  May take some time as I'll have to update the clients but won't be a huge change.

Also, be on a lookout for the redux library supporting this. I'll add it there as well.

thanks for your review 😄

File: store/sqlstore/upgrade.go (hunk 1 above)
Comment (jwilander) [reply store/sqlstore/upgrade.go]: >@jwilander, 150 refers to the closest maximum length with some padding (maybe like 6 char additional padding)

Any objections to bumping it to 256 just to be safe? It's a varchar so that space won't be taken up if it's not used

>Could you tell me what's the difference between varchar(150) and table.ColMap("Position").SetMaxSize(128)

Essentially there is no difference, but it needs to get added at the line I linked as well since that bit of code is for new installs. If we don't specify there it will default to something else

File: utils/config.go (hunk 2 above)
Comment (jwilander) [reply utils/config.go]: Sounds good :)

File: model/config.go (hunk 6 above)
Comment (jwilander) [reply model/config.go]: I'm with @miguelespinoza on this, I think it's good to have it configurable if needed

**Comment by miguelespinoza** [comment]
@jwilander, @lieut-data, @hmhealey.  I've made the changes requested 😄 

For reference, here's the redux counterpart: https://github.com/mattermost/mattermost-redux/pull/423
Currently, waiting for redux library to get merged so I can push latest changes for webapp and mobile.

Let me know if there's anything else on my part. I'm curious from a release standpoint which version would this feature more than likely fit.

File: api4/timezone.go [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+	supportedTimezones := c.App.Config().SupportedTimezones
+
+	if len(supportedTimezones) == 0 {
+		emptyTimezones := make([]string, 0)
+		w.Write([]byte(model.TimezonesToJson(emptyTimezones)))
+		return
+	}
+
+	w.Write([]byte(model.TimezonesToJson(supportedTimezones)))
+}
```
Comment (jwilander) [inline api4/timezone.go]: Could we throw this under the `/api/v4/system` route? I think it fits there better than having it's own root path

**Comment by jwilander** [comment]
Thanks @miguelespinoza ! Just that one small comment and it looks good to me.

The cut-off for merging major features for 4.9 is on March 27th, so assuming this goes in before that then it'll make it for 4.9

File: store/sqlstore/upgrade.go [hunk 8]
```diff
@@ -352,6 +353,13 @@ func UpgradeDatabaseToVersion47(sqlStore SqlStore) {
 		sqlStore.AlterColumnTypeIfExists("OAuthAuthData", "State", "varchar(1024)", "varchar(1024)")
 		sqlStore.RemoveColumnIfExists("ChannelMemberHistory", "Email")
 		sqlStore.RemoveColumnIfExists("ChannelMemberHistory", "Username")
+
+		defaultTimezone := model.DefaultUserTimezone()
+		defaultTimezoneValue, err := json.Marshal(defaultTimezone)
+		if err != nil {
+			l4g.Critical(err)
+		}
+		sqlStore.CreateColumnIfNotExists("Users", "Timezone", "varchar(256)", "varchar(256)", string(defaultTimezoneValue))
```
Comment (jwilander) [inline store/sqlstore/upgrade.go]: One last change, this needs to be moved down into the upgrade function for going to 4.9

File: model/config.go (hunk 6 above)
Comment (hmhealey) [reply model/config.go]: Sounds good

Comment (hmhealey) [reply model/config.go]: Are these going to get written back to the config.json when that gets saved?

File: utils/config.go [hunk 9]
```diff
@@ -278,16 +278,13 @@ func EnsureConfigFile(fileName string) (string, error) {
 	return "", fmt.Errorf("no config file found")
 }
 
-// LoadConfig will try to search around for the corresponding config file.  It will search
-// /tmp/fileName then attempt ./config/fileName, then ../config/fileName and last it will look at
-// fileName.
-func LoadConfig(fileName string) (config *model.Config, configPath string, appErr *model.AppError) {
+func loadConfigFile(fileName string) (config *model.Config, configPath string, appErr *model.AppError) {
```
Comment (hmhealey) [inline utils/config.go]: I actually just realized a potential problem here. What happens if the user specifies the path to their config.json using an environment variable and we try to load the timezones.json? This affects any GitLab Omnibus installs in particular.

It might be a bit hokey, but we might need to add a config setting for the path to the timezones file so that it can be loaded properly. Something like `TimezoneSettings.AvailableTimezonesPath` that, if empty, defaults to doing what it does currently?

Comment (miguelespinoza) [reply utils/config.go]: > Are these going to get written back to the config.json when that gets saved?

Normally when is it saved back to the config.json file?
I noticed admin can save the config, so I can add some logic to filter out `SupportedTimezones` in order to not update `config.json` with those values


> It might be a bit hokey, but we might need to add a config setting for the path to the timezones file so that it can be loaded properly. Something like TimezoneSettings.AvailableTimezonesPath that, if empty, defaults to doing what it does currently?

I can see this being useful. I can add that in 👍

**Comment by miguelespinoza** [comment]
Last changes made.

* Move db upgrade code to v4.9
* Filter out SupportedTimezones from getting saved to `config.json`
* Add TimezonesSettings.SupportedTimezonesPath for users like GitLabs Omnibus

**Comment by miguelespinoza** [comment]
Thanks @jwilander, just one last change.  I removed the `EnableTimezoneSelection` flag from the config based on @jasonblais requests 😄

**Comment by jwilander** [comment]
@hmhealey and @lieut-data if you could take another look that'd be great

**Comment by lieut-data** [comment]
👀 now.

File: api4/system.go [hunk 10]
```diff
@@ -378,6 +380,18 @@ func getAnalytics(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.Write([]byte(rows.ToJson()))
 }
 
+func getSupportedTimezones(c *Context, w http.ResponseWriter, r *http.Request) {
+	supportedTimezones := c.App.Config().SupportedTimezones
+
+	if len(supportedTimezones) == 0 {
```
Comment (lieut-data) [inline api4/system.go]: nit: is this block doing anything that line 392 wouldn't do already?

File: api4/system_test.go [hunk 11]
```diff
@@ -527,3 +527,26 @@ func TestS3TestConnection(t *testing.T) {
...
+	defer th.TearDown()
+	Client := th.Client
+
+	supportedTimezonesFromConfig := th.App.Config().SupportedTimezones
+	supportedTimezones, resp := Client.GetSupportedTimezone()
+
+	CheckNoError(t, resp)
+	for _, timezone := range supportedTimezones {
+		found := false
+		for _, configTimezone := range supportedTimezonesFromConfig {
```
Comment (lieut-data) [inline api4/system_test.go]: nit: This does indeed verify that the client returns only configurations that are supported, but doesn't verify that it returns all of them. Would it be simpler just to use `assert.Equal(t, supportedTimezonesFromConfig, supportedTimezones)`, given that they are going to already be in the same order?

File: utils/config.go [hunk 12]
```diff
@@ -144,7 +144,7 @@ func GetLogFileLocation(fileLocation string) string {
 }
 
 func SaveConfig(fileName string, config *model.Config) *model.AppError {
-	b, err := json.MarshalIndent(config, "", "    ")
+	b, err := json.MarshalIndent(config.SaveClone(), "", "    ")
```
Comment (lieut-data) [inline utils/config.go]: It took me a while before I realized what `SaveClone` actually did. I'm worried about the presence of such a function. Is there a reason we don't just add `json:"-"` to the `SupportedTimezones` field on the config to prevent its serialization? Or are we serializing the full configuration in other contexts?

File: utils/config.go [hunk 13]
```diff
@@ -278,16 +278,13 @@ func EnsureConfigFile(fileName string) (string, error) {
...
-// fileName.
-func LoadConfig(fileName string) (config *model.Config, configPath string, appErr *model.AppError) {
+func loadConfigFile(fileName string) (config *model.Config, configPath string, appErr *model.AppError) {
 	if fileName != filepath.Base(fileName) {
 		configPath = fileName
 	} else {
 		if path, err := EnsureConfigFile(fileName); err != nil {
 			appErr = model.NewAppError("LoadConfig", "utils.config.load_config.opening.panic", map[string]interface{}{"Filename": fileName, "Error": err.Error()}, "", 0)
-			return
+			return nil, "", appErr
```
Comment (lieut-data) [inline utils/config.go]: If we're not using the named parameters in the function signature anymore, should we remove them?

File: utils/config.go [hunk 14]
```diff
@@ -296,7 +293,19 @@ func LoadConfig(fileName string) (config *model.Config, configPath string, appEr
...
+		return nil, "", appErr
+	}
+
+	return config, configPath, nil
+}
+
+// LoadConfig will try to search around for the corresponding config file.  It will search
+// /tmp/fileName then attempt ./config/fileName, then ../config/fileName and last it will look at
+// fileName.
+func LoadConfig(fileName string) (config *model.Config, configPath string, appErr *model.AppError) {
```
Comment (lieut-data) [inline utils/config.go]: Similar comment as above: remove the named parameters?

File: utils/config.go (hunk 12 above)
Comment (miguelespinoza) [reply utils/config.go]: I've tried 2 other approaches

1. json:"-"
This avoids Deserialization and Serialization. Therefore, pulling `timezones.json` wouldn't work.

2. SupportedTimezones implementing the json.Marshal interface.

If I'm recalling this perfectly:
```
func (SupportedTimezones) MarshalJSON() ([]byte, err) {
   return []byte("[]",) nil
}
```

Works but you end up with this in `config.json`
```
"SupportedTimezones": []
```

Not pretty 😖 

Therefore, creating a new struct that serializes only what we need was the best option.  Actually the preferred solution in the community.  It get's the job done, and avoids saving extraneous data

File: utils/config.go (hunk 13 above)
Comment (miguelespinoza) [reply utils/config.go]: Ah, I missed the purpose of named parameters.  Been some time I've done Go dev, I may have missed some best practices 😓 

I like the documentation's explanation of having named parameters as a way of documenting your code
https://golang.org/doc/effective_go.html#named-results
> The names are not mandatory but they can make code shorter and clearer: they're documentation. If we name the results of nextInt it becomes obvious which returned int is which.

Either way, this does avoid us from initializing the variables so it's not completely pointless.
If we want to take full advantage of the named parameter feature, do you prefer returning without explicitly stating the results?

Basically `return` instead of `return nil, "", appErr` <- that can get accomplished by the named result parameters feature

Comment (miguelespinoza) [reply utils/config.go]: I would prefer keeping the named parameters. that would provide a cleaner solution

I'll revet back to only `returning`

File: api4/system.go (hunk 10 above)
Comment (miguelespinoza) [reply api4/system.go]: @lieut-data my concern was if supportedTimezones was nil, it wouldn't have the same effect as line 392.

I believed that serializing `nil` would not provide an empty JSON struct of "[]"?

Being that Go is typed, would it respect `type Supportedtimezones []string` and return the zero value for that which would be `"[]"`?

The goal is to have consistency, **always** return an array as a response.  I've had issues with other APIs where it would return null or {} to clients which expected [].

This would cause inexplicable side-effects on clients that were typed such on Android w/ Java.

Just my experience, I am very aware of Go's type system.  It could very well be that Go takes care of that for you and I did not know 😞

Comment (lieut-data) [reply api4/system.go]: Ah, my mistake -- you're right that `nil` wouldn't be handled the same, and I agree it would be best to always return an array. Perhaps just check for `!= nil` to emphasize the intent, since a a non-nil but empty `supportedTimezones` should encode the same way.

File: utils/config.go (hunk 12 above)
Comment (lieut-data) [reply utils/config.go]: Ah, I understand now: `LoadTimezoneConfig` is deserializing into a `Config` data structure, vs. serializing into a structure of it's own. 

I agree that creating a new struct for custom serialization is often the right idea, but in this case it seems forced. `SupportedTimezones` is made to be part of the config structure but then excluded in almost every way. It has its own `LoadTimezoneConfig` that doesn't actually load the `TimezoneSettings` part of the `Config`, just the `SupportedTimezones`.

Have we considered making `SupportedTimezones` not part of the `Config`, but a sibling on the `App`? I'm 1/5 on that approach.

Or, what if `LoadTimezoneConfig` was written to just return a `SupportedTimezones` struct instead of a `Config` object, allowing `LoadConfig` to basically work the same way, but then trivially excluding the `SupportedTimezones` struct with the `json:"-"` when serializing to the client? Seems like the best of all worlds.

Apart from the modelling concerns above, I think my concern about the `SaveClone` method stems from the fact that:
1) It's not obvious that this function only excludes a single struct from the Config (you have to scan to ask, "What's missing?")
2) It seems easy to forget to update this method when adding a new configuration field that is meant to be saved.

My apologies for not providing this feedback earlier -- I hadn't been following the conversation on this pull request closely until @jwilander kindly poked me :)

File: utils/config.go (hunk 13 above)
Comment (lieut-data) [reply utils/config.go]: I know there's lots of strong opinions on named parameters, so I'm 0/5 on this at the moment, but to share from https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters:

>  Don't name result parameters just to avoid declaring a var inside the function; that trades off a minor implementation brevity at the cost of unnecessary API verbosity.

and 

> Naked returns are okay if the function is a handful of lines. Once it's a medium sized function, be explicit with your return values. Corollary: it's not worth it to name result parameters just because it enables you to use naked returns. Clarity of docs is always more important than saving a line or two in your function.

File: api4/system.go (hunk 10 above)
Comment (miguelespinoza) [reply api4/system.go]: Right, I had to step out so I wasn't able to post this example earlier: https://play.golang.org/p/qPUivJTSwYt

The reason I didn't explicitly check for `!= nil` was because len handles that.
https://golang.org/pkg/builtin/#len
> if v is nil, len(v) is zero.

File: utils/config.go (hunk 12 above)
Comment (miguelespinoza) [reply utils/config.go]: No problem, it can be hard to understand the issue without context.

Could you bear with me for a second, I quickly gave your approach a shot.

1. `SupportedTimezones    SupportedTimezones json:"-"` in `model/config.go  type Config`

2. Here's the `LoadTimezones` function
```
func LoadTimezones(fileName string) (timezones model.SupportedTimezones, configPath string, appErr *model.AppError) {
	config, _, _ := loadConfigFile(fileName)
	return config.SupportedTimezones, "", nil
}
```

3. And in `app/config.go` the transfer process
```
supportedTimezones, _, err := utils.LoadTimezones(timezonePath)
if err != nil {
	return err
}

cfg.SupportedTimezones = supportedTimezones
```

This is where it gets rather odd.
When I call the `system/timezones` endpoint.

I get no timezones.  The array is empty.
If I remove `json:"="` it works.

My assumptions about `json` are wrong.  I thought it only affects serialize processing.  Am I wrong? I may be missing something

Comment (hmhealey) [reply utils/config.go]: Adding the `json:'-'` struct tag to the Config struct will affect any serialization and deserialization of that, but I don't think it should affect serializing the value of the field. 

Is this what you were thinking, Jesse? https://gist.github.com/hmhealey/ff79c43a3fce0dc1d1a87e1a35d4f399

Comment (lieut-data) [reply utils/config.go]: @hmhealey, yeah that's the spirit of what I had hoped would simplify the code for @miguelespinoza.

The reason `timezones` is empty above is because `LoadTimezones` in your example still depends on deserializing the Config structure, in which the `SupportedTimezones` field is now marked as ignored. If we agree this is the right direction, I'm hoping we can decouple the supported timezones from config as much as possible, landing on something less tied to the existing config parsing, ala:

```go
func LoadTimezones(fileName string) (model.SupportedTimezones, *model.AppError) {
          var supportedTimezones model.SupportedTimezones

          if timezoneFile := FindConfigFile(fileName); timezoneFile == "" {
                  return nil, model.NewAppError("LoadTimezones", "no.such.file", map[string]interface{}{"Filename": fileName}, "", 0)
          } else if raw, err := ioutil.ReadFile(timezoneFile); err != nil {
                  return nil, model.NewAppError("LoadTimezones", "failed.to.open.file", map[string]interface{}{"Filename": fileName, "Error": err.Error()}, "", 0)
          } else if err := json.Unmarshal(raw, &supportedTimezones); err != nil {
                  return nil, model.NewAppError("LoadTimezones", "not.json", map[string]interface{}{"Filename": fileName, "Error": err.Error()}, "", 0)
          } else {
                  return supportedTimezones, nil

          }
  }
```

Of course, this also implies flattening `timezones.json`:
```
[
    "Africa/Abidjan",
    "Africa/Accra",
    "Africa/Addis_Ababa",
    "Africa/Algiers",
    "Africa/Asmara",
    "Africa/Asmera",
    "Africa/Bamako",
    "Africa/Bangui",
    "Africa/Banjul",
...
]
```

Comment (miguelespinoza) [reply utils/config.go]: @hmhealey and @lieut-data, I see, thanks. I can take the approach of decoupling the timezone from the config.

Just for some guidance would this live in `app.go`, following the same pattern as config or license?
Basically, `timezone       atomic.Value`
Or is there a simpler path for supporting timezones.  The goal is to call `systems/timezones` which was currently in this format

```
func getSupportedTimezones(c *Context, w http.ResponseWriter, r *http.Request) {
	supportedTimezones := c.App.Config().SupportedTimezones

	if len(supportedTimezones) == 0 {
		emptyTimezones := make([]string, 0)
		w.Write([]byte(model.TimezonesToJson(emptyTimezones)))
		return
	}

	w.Write([]byte(model.TimezonesToJson(supportedTimezones)))
}
```

I imagine decoupling the timezone would look something like
`supportedTimezones := c.App.Timezones()`

Comment (miguelespinoza) [reply utils/config.go]: This would be the last thing, hopefully, once we can come to an agreement I'll get this in and we can get this ready for 4.9 😄

Comment (hmhealey) [reply utils/config.go]: That sounds like it would work well. I think the problems we're running into here are mostly caused by having the timezones as a kind-of-but-not-really part of the config. Separating them out should fix that

**Comment by miguelespinoza** [comment]
# Timezone support for Mattermost
* webapp: https://github.com/mattermost/mattermost-webapp/pull/838
* mobile: https://github.com/mattermost/mattermost-mobile/pull/1503

Above PRs have changes that removed JS Date API for properly supporting timezones.

A proposed solution from Koxen
> I guess what you can do for your PR is steal the timezone offsets from moment.js and use that to modify timestamp
> https://momentjs.com/timezone/docs/#/zone-object/offset/

I would ideally stay away from having to import a timezone offset database for both web and mobile.

I decided to work off of the existing server solution.  Each `User` contains a `Timezone` field which currently has

```
{
	useAutomaticTimezone: true,
	automaticTimezone: "America/New_York"
	manualTimezone: ""
}
```

My idea is to add a new property `timezoneOffset` that gets calculated every time the timezone changes either automatic or manually.

Updated `Timezones` field:
```
{
	useAutomaticTimezone: true,
	automaticTimezone: "America/New_York"
	manualTimezone: ""
	timezoneOffset: -5 // in hrs, but actually converted to ms
}
```

Works perfectly fine until DST comes along.  Say the user never travels therefore has “America/New_York” all the time.  There’s no need to update the timezone ever.

Last week my timezone offset would have been UTC -5 well now it’s UTC -4.
Great, now all my new posts timestamps are 1 hour into the future!

We would have to update the timezoneOffset every time a user begins a session, just to ensure we don’t mistakenly show the wrong offset.  
I’m against this approach, it’s an unnecessary process to have to check if the timezone needs to be updated because of DST implications.

Saying all this, we are in this current state because we avoid using timezone libraries for both web and mobile,  we are reinventing the wheel for an issue that can be solved with:
```
date.toLocaleString()
```

If removing the above code, increases performance for mobile.   We have other aspects to truly help with performance
	1. Refactor JS Navigation drawer in favor of a native solution
	2. Improve Layout Hierarchy to stop Overdraw (Android specific)

We can drastically improve performance in these areas which would make low hanging fruits like FormatTime speed obsolete.

At this point, we are confronted with this issue: **convenience vs performance.**

Pros and cons of removing `date.toLocaleString()`?
	* PROS: Improve performance by an unknown metric
	* CONS: Simple logic to calculate hours and minutes, which means we’ll need to add new logic for things like offset calculation or DST
	* CONS: include additional property `timezoneOffset` (server)
	* CONS: include additional logic to check/recheck if `timezoneOffset` is valid (server)

@jwilander, @lieut-data @enahum @hmhealey 
Let me know any suggestions, I would like to get your thoughts before I move along with any solution

**Comment by hmhealey** [comment]
If the removal of `toLocaleString()` is causing us this many problems, I say that we revert it. I ran some speed tests earlier, and at least on Windows, doing it manually is significantly faster, but the "slow" case is still not slow enough to be notable.

image

The "Manual" case is how we're doing it now, and I believe the previous behaviour was using the "DateTimeFormat (saved)" case. So while it's almost 100x faster now, that's on the scale of a handful of microseconds vs whatever is smaller than that. We don't render dates and timestamps enough in the app for that to have any noticeable effect. Even if it's the first case that we're actually experiencing, it would be easy enough to improve that drastically by switching to the third one.

**Comment by enahum** [comment]
I’m with harrison here, if it isn’t really noticeable we can just revert it, it seems that we have more cons than pros and then we should shift our focus to improve performance in other ways.

File: config/timezones.json [hunk 15]
```diff
@@ -0,0 +1,596 @@
+[
```
Comment (coreyhulen) [inline config/timezones.json]: What happens when this file does not exist?  Does it fall back to code?

We have a lot of automated installs that may need to change if this file is required.

Comment (miguelespinoza) [reply config/timezones.json]: There is no fallback.  If the file does not exist, the server cannot run without it.

It would be a similar issue if the `config.json` file did not exist.  Unless I'm missing something, I believe there's no fallback if `config.json` is missing. 

Quick test: move `config,default.json` into a `config/tmp` directory and run the server.  I get this error:

```
Error: LoadConfig: Error opening config file=config.json, err=no config file found,
[15:27:50 EDT 2018/03/19] [INFO] (github.com/mattermost/mattermost-server/app.(*App).Shutdown:206) Stopping Server...
[15:27:50 EDT 2018/03/19] [INFO] (github.com/mattermost/mattermost-server/app.(*App).HubStop:114) stopping websocket hub connections
[15:27:50 EDT 2018/03/19] [WARN] (github.com/mattermost/mattermost-server/app.(*App).HubStop:119) We appear to have already sent the stop checking for deadlocks command
[15:27:50 EDT 2018/03/19] [INFO] (github.com/mattermost/mattermost-server/app.(*App).Shutdown:225) Server stopped
[15:27:50 EDT 2018/03/19] [CRIT] (github.com/mattermost/mattermost-server/cmd/commands.runServer:64) LoadConfig: Error opening config file=config.json, err=no config file found,
```



> We have a lot of automated installs that may need to change if this file is required.

I'm curious what that setup looks like.   The `timezones.json` file should be packaged with the automated install script, is this handled by MM or is this a user-specific feature?

Is this what you are referring to? https://docs.mattermost.com/guides/orchestration.html?highlight=automated%20installs#deployment-solution-program-requirements


Let me know how I can help with this issue if there's anything I can do 😄 

Thanks for the catch @coreyhulen

Comment (miguelespinoza) [reply config/timezones.json]: @coreyhulen is there a way I can test automated installation?

Comment (coreyhulen) [reply config/timezones.json]: Yeah not sure it's comparable to config.json since that's integral to the system running.

Why not have it in code?  I doubt this file would ever change.  Or if you really want the file then have both.  So it will fallback to the code version when the file is missing similar to localization.

Not sure you can test all the various installers.  For example the one I pretty sure will fail would be Gitlab Onimbus integration.  They generate the config directory and file at a different location.

Comment (miguelespinoza) [reply config/timezones.json]: I agree it's not integral to the system running.  For some context, this is a discussion I had with @hmhealey earlier last week in this PR. 

> I actually just realized a potential problem here. What happens if the user specifies the path to their config.json using an environment variable and we try to load the timezones.json? This affects any GitLab Omnibus installs in particular since they like to store their files all over the place.
>
> It might be a bit hokey, but we might need to add a config setting for the path to the timezones file so that it can be loaded properly. Something like TimezoneSettings.AvailableTimezonesPath that, if empty, defaults to doing what it does currently?

@coreyhulen I made a change to support GitLab Omnibus users.  Would this suffice?

Comment (miguelespinoza) [reply config/timezones.json]: I left it in a json file in order for admins to configure if they desire.
Say they want to only support timezones in one region.

@dmeza had a suggestion that I believe would work fine.
Have a `default_timezones.json` and `timezones.json` file. Similar to `default and config.json`.
Would that work?
If `timezones.json` does not exists we recreate it from `default_timezones.json`

Comment (coreyhulen) [reply config/timezones.json]: I'm not certain either one of those works 100%.  `TimezoneSettings.AvailableTimezonesPath` helps but the gitlab installer still needs to change.  default_timezones.json still has the same problem (it would need to exist in the dir).  Unless you meant `default_timezones.json` is in a go string that is deserialized when the proper file is missing.

Comment (miguelespinoza) [reply config/timezones.json]: > Unless you meant default_timezones.json is in a go string that is deserialized when the proper file is missing.

The idea would be the same as `config/default.json` and `config.json`.
If `config.json` does not exists, fallback to `default.json`.

I was not thinking about a go string per se.

I understand your concern and I feel it's unnecessary to update the automated install process.
I'll add some logic to default to a list of supported timezones in go code.  That should be a succinct solution.

Comment (coreyhulen) [reply config/timezones.json]: :+1:

**Comment by miguelespinoza** [comment]
Thanks everyone for the dev review 😄 

@jwilander or @hmhealey Can this get merged in order to test both web and mobile PRs:
https://github.com/mattermost/mattermost-webapp/pull/696
https://github.com/mattermost/mattermost-mobile/pull/1456
